### ISSUE TITLE
🧪 Scientific Computing & Output Quality: NumPy/SciPy Prompting, Notebook Export, Rich Tables, Plot Themes, PDF Reports [10 sub-features]

### ISSUE THREAD

**Issue by haseeb-heaven** [haseeb-heaven/code-interpreter#223]
# 🧪 Scientific Computing & Output Quality

This issue covers everything a researcher, data scientist, or academic needs beyond basic pandas/matplotlib — scientific computing libraries, publication-quality output, Jupyter notebook export, and the output formatting that makes results actually usable. Combines 10 sub-features.

---

## Sub-feature 1: Scientific Computing System Prompt Layer

**What's missing:** When a scientist asks *"fit a Gaussian to this data"* or *"run a t-test between these two groups"*, the LLM generates generic Python instead of proper scipy/statsmodels/sklearn code. There's no scientific computing context in the system prompt.

**Fix — add `--science` flag and `libs/prompts/science_prompt.py`:**
```python
SCIENCE_SYSTEM_PROMPT = """
You are a scientific computing assistant. When performing data analysis:

- Statistics: use scipy.stats for hypothesis tests (t-test, ANOVA, chi-square, Mann-Whitney)
- Regression: use sklearn or statsmodels, always report R², p-values, confidence intervals
- Signal processing: use scipy.signal for filtering, FFT, spectral analysis
- Optimization: use scipy.optimize for curve fitting and minimization
- Linear algebra: use numpy.linalg, never implement from scratch
- Always report: sample size (n), effect size, p-value, confidence interval for any statistical test
- Plots: use seaborn for statistical plots (violin, box, swarm, regression), matplotlib for everything else
- Never round p-values — report full precision (e.g., p=0.0023, not p<0.05)
- Always set random seeds for reproducible results: np.random.seed(42)
"""
```

Activate with `--science` flag or auto-detect when prompt contains words like: "p-value", "correlation", "regression", "hypothesis", "significance", "confidence interval", "ANOVA", "distribution fit".

---

## Sub-feature 2: Jupyter Notebook Export

**What's missing:** Researchers want `.ipynb` files they can share with colleagues, submit to journals, or continue working in. Generated code disappears after execution with no way to get a notebook.

**Fix — `libs/output/notebook_exporter.py`:**
```python
import json
from datetime import datetime
from pathlib import Path

def export_to_notebook(
    cells: list[dict],   # [{"type": "code"|"markdown", "source": str, "output": str}]
    output_path: str = None,
    title: str = "Code Interpreter Session"
) -> str:
    """Convert a session's code+output history to a .ipynb file."""
    nb = {
        "nbformat": 4,
        "nbformat_minor": 5,
        "metadata": {
            "kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"},
            "language_info": {"name": "python", "version": "3.10.0"}
        },
        "cells": []
    }

    # Title cell
    nb["cells"].append({
        "cell_type": "markdown",
        "metadata": {},
        "source": [f"# {title}\n", f"*Generated by Code Interpreter on {datetime.now().strftime('%Y-%m-%d %H:%M')}*"]
    })

    for cell in cells:
        if cell["type"] == "markdown":
            nb["cells"].append({
                "cell_type": "markdown",
                "metadata": {},
                "source": cell["source"].splitlines(keepends=True)
            })
        else:
            outputs = []
            if cell.get("output"):
                outputs.append({
                    "output_type": "stream",
                    "name": "stdout",
                    "text": cell["output"].splitlines(keepends=True)
                })
            nb["cells"].append({
                "cell_type": "code",
                "execution_count": None,
                "metadata": {},
                "source": cell["source"].splitlines(keepends=True),
                "outputs": outputs
            })

    path = output_path or f"session_{datetime.now().strftime('%Y%m%d_%H%M%S')}.ipynb"
    Path(path).write_text(json.dumps(nb, indent=2))
    return path
```

**Add `/notebook` REPL command:**
```
/notebook         ← export current session as .ipynb
/notebook open    ← export + open in Jupyter if installed
/notebook save my_analysis.ipynb
```

The session tracker (`libs/data/session_data.py` from Issue #222) records each code+output pair to enable this.

---

## Sub-feature 3: Rich Terminal Output Formatting

**What's missing:** `rich` is already in `requirements.txt` but output is plain text. DataFrames print as ugly monospace. Numbers have no formatting. Errors are raw tracebacks. ChatGPT Code Interpreter shows nicely formatted tables, syntax-highlighted code, and clean error summaries.

**Fix — `libs/output/rich_formatter.py`:**
```python
from rich.console import Console
from rich.table import Table
from rich.syntax import Syntax
from rich.panel import Panel
from rich import box
import pandas as pd

console = Console()

def print_dataframe(df: pd.DataFrame, title: str = "", max_rows: int = 20):
    """Print a pandas DataFrame as a beautiful rich table."""
    table = Table(
        title=title,
        box=box.ROUNDED,
        show_header=True,
        header_style="bold cyan",
        row_styles=["none", "dim"]
    )
    for col in df.columns:
        # Right-align numeric columns
        justify = "right" if pd.api.types.is_numeric_dtype(df[col]) else "left"
        table.add_column(str(col), justify=justify)
    for _, row in df.head(max_rows).iterrows():
        table.add_row(*[_fmt_cell(v) for v in row])
    if len(df) > max_rows:
        table.caption = f"... {len(df) - max_rows} more rows"
    console.print(table)

def print_stats(stats: dict):
    """Print a key-value stats summary in a panel."""
    table = Table(box=box.SIMPLE, show_header=False, padding=(0,1))
    table.add_column("Metric", style="bold")
    table.add_column("Value", style="cyan")
    for k, v in stats.items():
        table.add_row(str(k), str(v))
    console.print(Panel(table, title="📊 Results", border_style="green"))

def print_code(code: str, language: str = "python"):
    console.print(Syntax(code, language, theme="monokai", line_numbers=True))

def print_error(error: str):
    console.print(Panel(error, title="❌ Error", border_style="red", style="red"))

def _fmt_cell(v) -> str:
    if isinstance(v, float):
        return f"{v:,.4f}" if abs(v) < 1e6 else f"{v:.4e}"
    if isinstance(v, int):
        return f"{v:,}"
    return str(v)
```

**Auto-detect DataFrames in output:** After code execution, if the last expression is a DataFrame, auto-print it with `print_dataframe()` instead of the default repr.

---

## Sub-feature 4: Publication-Quality Plot Themes

**What's missing:** Default matplotlib plots look like student homework. Scientists need publication-ready figures for papers, presentations, and reports. `seaborn` style exists but there's no guidance in prompts.

**Fix — add `--plot-theme` flag:**
```bash
python interpreter.py --plot-theme paper "plot regression of height vs weight"
```

**Available themes:**
```python
PLOT_THEMES = {
    "paper": {
        "style": "seaborn-v0_8-paper",
        "context": "paper",
        "font_scale": 1.2,
        "rc": {"figure.dpi": 300, "savefig.dpi": 300, "font.family": "serif"}
    },
    "talk": {
        "style": "seaborn-v0_8-talk",
        "context": "talk",
        "font_scale": 1.4,
        "rc": {"figure.dpi": 150}
    },
    "dark": {
        "style": "dark_background",
        "rc": {"figure.dpi": 150, "axes.facecolor": "#1a1a2e"}
    },
    "minimal": {
        "style": "seaborn-v0_8-whitegrid",
        "rc": {"figure.dpi": 150, "axes.spines.top": False, "axes.spines.right": False}
    }
}
```

Inject theme setup code at the top of every generated matplotlib script. Also inject: `plt.rcParams['figure.figsize'] = (10, 6)` by default (current default 6.4×4.8 is too small for analysis work).

---

## Sub-feature 5: PDF Report Generation

**What's missing:** Researchers and analysts need to share results. There's no report generation. The `/export report` in Issue #222 covers HTML — this adds PDF.

**Fix — extend `libs/output/exporter.py` with PDF support:**

Use `reportlab` (pure Python, no system dependencies) to generate a clean PDF report:

```python
def export_pdf_report(session, charts: list[Path], output_path: str = None) -> str:
    """
    Generate a PDF report containing:
    - Title page with dataset name + date
    - Dataset summary (shape, schema, null counts)
    - All generated charts embedded as images
    - Statistical summary tables
    - Key findings paragraph (LLM-generated summary)
    """
    from reportlab.lib.pagesizes import letter
    from reportlab.platypus import SimpleDocTemplate, Paragraph, Image, Spacer, Table
    from reportlab.lib.styles import getSampleStyleSheet
    # ... build PDF with reportlab
```

**Add `/export pdf` REPL command** and `--report` CLI flag:
```bash
python interpreter.py --eda sales.csv --report
# Runs full EDA + generates sales_report_20260712.pdf
```

---

## Sub-feature 6: Machine Learning Shortcuts

**What's missing:** The most common ML workflow (load data → train model → evaluate → predict) requires a lot of boilerplate. Scientists waste time on sklearn setup instead of the actual task.

**Fix — add `/ml` REPL command with sub-commands:**
```
/ml classify [target_column]   ← auto-select + train best classifier, show accuracy/F1
/ml regress [target_column]    ← auto-select + train best regressor, show R²/RMSE
/ml cluster [n_clusters]       ← K-Means clustering, show cluster profiles
/ml forecast [date_col] [val]  ← time series forecast using statsmodels ARIMA
/ml feature-importance         ← show feature importance from last trained model
```

Each command generates a full sklearn pipeline with:
- Train/test split (80/20)
- StandardScaler for numeric features
- OneHotEncoder for categorical features  
- Cross-validation (5-fold)
- Performance metrics with interpretation
- Feature importance visualization

---

## Sub-feature 7: Image/File Output Preview in Terminal

**What's missing:** When a chart is saved, the terminal just shows the file path. On terminals that support it (iTerm2, Kitty, Windows Terminal with sixel), the image should display inline.

**Fix — `libs/output/terminal_preview.py`:**
```python
import sys, os, subprocess
from pathlib import Path

def try_inline_preview(image_path: Path):
    """Try to display image inline in terminal if supported."""
    term = os.environ.get("TERM", "")
    term_program = os.environ.get("TERM_PROGRAM", "")

    # iTerm2 inline images protocol
    if term_program == "iTerm.app":
        _iterm2_display(image_path)
        return

    # Kitty terminal graphics protocol
    if term == "xterm-kitty":
        subprocess.run(["kitty", "+kitten", "icat", str(image_path)], check=False)
        return

    # Fallback: just print the path clearly
    print(f"📊 Chart saved → {image_path}")
    print(f"   Open with: open {image_path}  (mac) | xdg-open {image_path}  (linux)")

def _iterm2_display(path: Path):
    import base64
    data = base64.b64encode(path.read_bytes()).decode()
    size = path.stat().st_size
    print(f"\033]1337;File=inline=1;size={size};width=80%;height=auto:{data}\a")
```

---

## Sub-feature 8: Multi-File Batch Analysis

**What's missing:** Researchers often have many files — experiment1.csv, experiment2.csv, etc. There's no batch mode.

**Fix — add glob pattern support to `--file` flag:**
```bash
# Analyze all CSV files in a folder
python interpreter.py -f "./data/*.csv" "compare sales across all months"

# Multiple specific files
python interpreter.py -f q1.csv q2.csv q3.csv "show quarterly trend"
```

**In `libs/context/file_context.py`** (from Issue #221), expand `build_file_context()` to:
1. Accept glob patterns and expand them
2. When multiple files have the same schema, note they can be `pd.concat()`'d
3. When files have different schemas, describe each separately
4. Add a suggested code snippet at the bottom showing how to load all files

---

## Sub-feature 9: Code Output Capture & Display Improvement

**What's missing:** Long output (printing 1000 rows, verbose training logs) floods the terminal and obscures the result. There's no truncation, no paging, no summary.

**Fix — in `libs/execution/executor.py`:**
```python
MAX_OUTPUT_LINES = 50
MAX_OUTPUT_CHARS = 5000

def format_output(raw_output: str) -> str:
    lines = raw_output.splitlines()
    if len(lines) > MAX_OUTPUT_LINES:
        kept = lines[:20] + [
            f"\n... [{len(lines) - 40} lines truncated] ...\n"
        ] + lines[-20:]
        return "\n".join(kept)
    if len(raw_output) > MAX_OUTPUT_CHARS:
        return raw_output[:MAX_OUTPUT_CHARS] + f"\n... [truncated, {len(raw_output)} total chars]"
    return raw_output
```

Also add `/output full` REPL command to print the last full untruncated output to a temp file and open it in the system pager (`less`).

---

## Sub-feature 10: Requirements Auto-Installer

**What's missing:** Generated code often imports libraries not in `requirements.txt` — `scikit-learn`, `seaborn`, `statsmodels`, `scipy`, `openpyxl`, `duckdb`, etc. Right now this causes an `ImportError` that burns a repair attempt. The tool should just install missing packages.

**Fix — in `libs/execution/executor.py`, pre-execution import scan:**
```python
import ast, sys, subprocess

STD_SCIENCE_PACKAGES = {
    "sklearn": "scikit-learn",
    "cv2": "opencv-python",
    "PIL": "Pillow",
    "seaborn": "seaborn",
    "statsmodels": "statsmodels",
    "scipy": "scipy",
    "openpyxl": "openpyxl",
    "duckdb": "duckdb",
    "pyarrow": "pyarrow",
    "reportlab": "reportlab",
    "tabulate": "tabulate",
    "xlrd": "xlrd",
}

def auto_install_missing(code: str) -> list[str]:
    """Parse imports from code, install any missing packages. Returns list of installed."""
    try:
        tree = ast.parse(code)
    except SyntaxError:
        return []

    imports = set()
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                imports.add(alias.name.split('.')[0])
        elif isinstance(node, ast.ImportFrom) and node.module:
            imports.add(node.module.split('.')[0])

    installed = []
    for imp in imports:
        try:
            __import__(imp)
        except ImportError:
            pkg = STD_SCIENCE_PACKAGES.get(imp, imp)
            print(f"📦 Installing missing package: {pkg}...")
            subprocess.run(
                [sys.executable, "-m", "pip", "install", "-q", pkg],
                check=True, capture_output=True
            )
            installed.append(pkg)
    return installed
```

Run `auto_install_missing(code)` before every execution. Print a one-line notice per package installed. This eliminates the most common class of repair failures.

---

## Files to Create / Modify

| File | Purpose |
|---|---|
| `libs/prompts/science_prompt.py` | Scientific computing system prompt |
| `libs/output/notebook_exporter.py` | Session → .ipynb export |
| `libs/output/rich_formatter.py` | Rich terminal tables + formatting |
| `libs/output/terminal_preview.py` | Inline image preview for supported terminals |
| `libs/execution/executor.py` | Add `auto_install_missing()`, output truncation |
| `interpreter.py` | Add flags: `--science`, `--plot-theme`, `--report` |
| `libs/repl_commands.py` | Add: `/notebook`, `/ml`, `/output`, `/templates` |
| `requirements.txt` | Add: `seaborn`, `scipy`, `scikit-learn`, `statsmodels`, `reportlab`, `openpyxl`, `pyarrow`, `tabulate` |
| `README.md` | Add "For Scientists" section with examples |

---

## The Result

After this issue, a researcher or data scientist can:
1. `python interpreter.py --science -f experiment.csv "run ANOVA across groups"` → proper statistical output with p-values, effect sizes
2. `/notebook` → entire session exported as a shareable .ipynb file
3. `python interpreter.py --eda data.csv --report` → publication-ready PDF report
4. `/ml classify target_col` → trained model with cross-validated accuracy in 30 seconds
5. Missing packages install automatically — zero `ImportError` failures
6. Charts saved at 300 DPI for paper submission with `--plot-theme paper`

This makes the tool the go-to for academic researchers, data scientists, and analysts who need real scientific computing — not just toy examples.

**Comment by haseeb-heaven** [comment]
Merged to develop.

Evidence: --science/--plot-theme/--report; /notebook /ml /output; PDF export; auto-install; truncation; glob attach. Tests: 10 OK.
